plugins/SpectrumAnalyzer/SaWaterfallView.cpp

```diff
@@ -75,89 +93,116 @@ void SaWaterfallView::paintEvent(QPaintEvent *event)
 	painter.setRenderHint(QPainter::Antialiasing, true);
 
 	// check if time labels need to be rebuilt
-	if ((float)m_processor->m_inBlockSize / m_processor->getSampleRate() != m_oldTimePerLine)
+	if (secondsPerLine() != m_oldSecondsPerLine || m_processor->waterfallHeight() != m_oldHeight)
 	{
 		m_timeTics = makeTimeTics();
-		m_oldTimePerLine = (float)m_processor->m_inBlockSize / m_processor->getSampleRate();
+		m_oldSecondsPerLine = secondsPerLine();
+		m_oldHeight = m_processor->waterfallHeight();
 	}
 
 	// print time labels
 	float pos = 0;
 	painter.setPen(QPen(m_controls->m_colorLabels, 1, Qt::SolidLine, Qt::RoundCap, Qt::BevelJoin));
 	for (auto & line: m_timeTics)
 	{
-		pos = timeToYPixel(line.first, displayBottom);
+		pos = timeToYPixel(line.first, m_displayHeight);
 		// align first and last label to the edge if needed, otherwise center them
 		if (line == m_timeTics.front() && pos < label_height / 2)
 		{
-			painter.drawText(displayLeft - label_width - margin, displayTop - 1,
+			painter.drawText(m_displayLeft - label_width - margin, m_displayTop - 1,
 							 label_width, label_height, Qt::AlignRight | Qt::AlignTop | Qt::TextDontClip,
 							 QString(line.second.c_str()));
-			painter.drawText(displayRight + margin, displayTop - 1,
+			painter.drawText(m_displayRight + margin, m_displayTop - 1,
 							 label_width, label_height, Qt::AlignLeft | Qt::AlignTop | Qt::TextDontClip,
 							 QString(line.second.c_str()));
 		}
-		else if (line == m_timeTics.back() && pos > displayBottom - label_height + 2)
+		else if (line == m_timeTics.back() && pos > m_displayBottom - label_height + 2)
 		{
-			painter.drawText(displayLeft - label_width - margin, displayBottom - label_height,
+			painter.drawText(m_displayLeft - label_width - margin, m_displayBottom - label_height,
 							 label_width, label_height, Qt::AlignRight | Qt::AlignBottom | Qt::TextDontClip,
 							 QString(line.second.c_str()));
-			painter.drawText(displayRight + margin, displayBottom - label_height + 2,
+			painter.drawText(m_displayRight + margin, m_displayBottom - label_height + 2,
 							 label_width, label_height, Qt::AlignLeft | Qt::AlignBottom | Qt::TextDontClip,
 							 QString(line.second.c_str()));
 		}
 		else
 		{
-			painter.drawText(displayLeft - label_width - margin, pos - label_height / 2,
+			painter.drawText(m_displayLeft - label_width - margin, pos - label_height / 2,
 							 label_width, label_height, Qt::AlignRight | Qt::AlignVCenter | Qt::TextDontClip,
 							 QString(line.second.c_str()));
-			painter.drawText(displayRight + margin, pos - label_height / 2,
+			painter.drawText(m_displayRight + margin, pos - label_height / 2,
 							 label_width, label_height, Qt::AlignLeft | Qt::AlignVCenter | Qt::TextDontClip,
 							 QString(line.second.c_str()));
 		}
 	}
 
 	// draw the spectrogram precomputed in SaProcessor
-	if (m_processor->m_waterfallNotEmpty)
+	if (m_processor->waterfallNotEmpty())
```

Comment:
I have a hard time imagining a situation where this would be a problem (first of all, the waterfall would have to store at least 2^16 or more lines), but compared to the waterfall drawing the performance hit of using an atomic is probably negligible, so I might as well do it just for the sake of actually using an atomic for the first time.. :D

Reply:
> the waterfall would have to store at least 2^16 or more lines

If an int would consist of multiple byte writes, the problem would be at 256. Though I'm not sure such architectures really exist.

> I might as well do it just for the sake of actually using an atomic for the first time.. :D

~~Good that you have not read the other comments yet~~ :smile: